### DISCUSSION TITLE
BE | Service에서 Service를 의존할까 Repository(Dao)를 의존할까 | 2022.05.24

### DISCUSSION THREAD

**Discussion by woowabrie** [woowacourse/retrospective#15] [General]
## Service 객체가 다른 Service를 의존하는 것과 Repository를 의존하는 것은 어떻게 다를까요?
- 미션을 하며 위와 같은 고민을 어떤 경우에 했는지 남겨주세요.
- 본인이 어떤 선택을 했는지 그 이유를 함께 적어주세요.

**Comment by progress0407** [comment]
객체지향적인 관점에서도 중요하지만 별개로 서비스 계층에서 하는 가장 큰 것 중 하나는 트랜잭션 원자성이라고 생각합니다.
R이 아닌 CUD의 명령형 트랜잭션의 경우... DB의 일관성이 깨질 수 있기 때문에 서비스 운영상에 치명적인 해를 입힐 수 있습니다.
서비스가 다른 서비스를 알거나 할 때는 이 @Transactional 처리를 할 때 코드를 추적하기 다소 어려울 수 있습니다. 되도록 하나의 서비스 계층을 갖거나... 순환 참조 등의 일을 피하여 계층형으로 (다른 서비스를 아우르는 서비스 계층을 하나 둠) 두면 좋을 것 같습니다.

객체지향적인 구조는 나중에라도 리펙터링을 통해서 추출할 수 있지만... 깨진 데이터를 원상태로 돌리는 것은 회사의 신뢰성 문제 등... 많은 문제를 야기할 것 같습니다... ㅠㅠ

**Reply by jayjaehunchoi** [reply]
Service를 특정 메서드에서 사용하기로 했다면 하나의 트랜잭션으로 보기로 한 것 아닌가요? 기본 트랜잭션 옵션을 사용한다면 트랜잭션이 전파될것같은데욥

**Reply by progress0407** [reply]
기본 정책으로... 부모로 트랜잭션이 합류된다 알 수는 있긴 하죠.. 다만 분산되어서 서로 트랜잭션 어노테이션을 확인해주어야 하니까 실수의 소지가 있어 보여요.

예를들어 한쪽 클래스는 클래스레벨이 read only transactional (그리고 메서드레벨에서 세부 설정)이고 다른 한쪽은 read-write transactional이면... 헷갈릴 수 있으니까요

**Comment by hyeonic** [comment]
서비스 계층에서 주는 이점들이 많다. 단순히 Dao를 반복해서 사용하게 되면 데이터 무결성을 지키기 위한 검증 들이 불필요하게 늘어나는 문제가 발생한다. 또한 서비스 계층을 의존하게 되면 트랜잭션에 대한 처리까지 보장할 수 있다. 만약 서비스 계층 간 순환참조가 일어나게 되면 그것은 잘못된 설계를 하고 있다는 것을 내포한다고 생각하고 적절히 책임을 이전하여 개선해야 한다고 판단한다.

**Comment by jayjaehunchoi** [comment]
비즈니스 제약사항의 사용이 필요하다면 (일반적으로 CUD)Service, 그렇지 않다면 Repository (일반적으로 Read)!

직접 상호 참조하는 일은 막으면 될 것 같고, 여러 객체를 타고 결국 서로를 알고 있다면 스프링 실행 자체가 안될텐데 그런 경우를 예방하는 방법을 사용하면 좋을 것 같아요~!

**Comment by jinyoungchoi95** [comment]
Repository를 의존하게 되면 Repository에서 가져온 다른 도메인들을 쉽게 사용할 수 있음.
따라서 가벼운 어플리케이션을 만든다면 그냥 Service가 다른 도메인의 Repository를 의존하여 사용하는 것이 오히려 더 개발하는데 용이하다고 생각함.

다만 어플리케이션의 규모가 커질수록 도메인도 커지고 도메인을 관리하는 Service도 많아짐. 이때 하나의 도메인 Service가 다양한 Repository를 관리하게 된다면 도메인 Service의 책임이 너무 무거워짐. 또한 해당 도메인을 관리하는 Service가 다른 도메인도 관리하게 되어 Service의 역할도 모호해짐.

따라서 하나의 Service가 하나의 Repository를 갖게하여 도메인을 관리하도록 하고 이들을 관리하는 상위 계층의 Service를 두어 트랜잭션관리를 하는 것을 생각하게 되었음.
순환 참조의 경우 순환 참조를 하게되는 상황자체를 막도록 리팩터링하는 것이 더 중요하며 순환참조를 사전에 확인하기 위해서 생성자 참조를 사용하는 이유도 있기 때문에 순환 참조가 발생하는 상황자체를 우려하지 않아도 된다고 생각함. 발생한다라면 잘못된 설계를 한 것이 아닌지 고민하면 해결가능하다고 생각함.

**Comment by HJ-Rich** [comment]
### 비슷한 고민이 복수의 레이어에 걸쳐 나타났어요
- 서비스가 서비스를 의존해도 될까?
- 레포지토리가 다른 레포지토리를 의존해도 될까?

<br> 

### 같은 레이어 내에서의 협력은 괜찮지 않을까요?
- 도메인 레이어 내에서 도메인 객체들은 적절한 역할 분배를 통해 서로 협력합니다.
- 따라서, 다른 레이어에서도 마찬가지로, 하나의 책임을 가진 객체가 다른 객체와 협력할 수 있다고 생각했어요.
- 혹은 의존이 안정적인 방향으로 향하는 경우도 괜찮다고 생각했어요. 가령 서비스는 서비스와 레포지토리 모두 의존 가능하다고 생각해요.

<br>

### 서비스 레이어는 약간 더 주의가 필요할 것 같아요
- 서비스 레이어에서는 다른 레이어와 특징적으로 구분되는 트랜잭션을 관리하는 역할이 있어요.
- 그로인해 서비스 레이어가 또다른 서비스 레이어를 의존할 경우, 트랜잭션 처리에 대해 주의가 필요할 것 같아요

<br>

### 서비스 레이어간 순환참조가 발생한다면?
- 순환 참조의 구조가 왜 발생했는지, 어떤 문제가 발생할지 살펴보는 게 우선일 것 같아요.
- 그러나 현재 순환참조가 발생한다는 사실 자체가 서비스 레이어끼리 의존을 하지 말아야 할 이유로는 충분치 않은 것 같아요.
- 순환 참조 구조를 개선하고, 서비스끼리 의존은 유지할 수도 있을 것 같아요.

**Comment by bugoverdose** [comment]
이번 미션에서는 서비스가 서비스를 참조할 일이 없었지만, 왜 하면 안된다고 생각하는지를 잘 모르겠어요. 

서비스가 서비스를 참조하면 순환참조가 될 수도 있다고 말하는데, 순환참조가 발생했다면 애초에 서비스들이 잘못 설계된게 아닌가요?

**Comment by Gyuchool** [comment]
### service 참조
 - 순환 참조가 발생할 가능성이 있다. 
 - A라는 서비스에서 B라는 서비스를 불러올 때, B의 레포지토리에서 B의 서비스를 만들어줘야하는데 이 때, A서비스를 위한 B서비스가 만들어지는 게 아닌가 라는 생각을 해볼 수도 있다.
 - 데이터가 분산되지 않고 관리하기 편할수있다.

### Repository 참조
- service -> repo 라는 단방향 계층으로 순환 참조같은 상황을 걱정하지 않아도 된다.
- 여러 서비스에서 데이터를 쉽게 불러올수있어서 유연하다.
- 데이터를 가져오는 로직이 여러 서비스에 분산되어서 관리하기 어려울 수 있다.

저는 작은 프로젝트의 경우엔 service에서 repo를 가져오도록 할 것 같아요. 데이터가 분산되기 작은 경우고 유연하게 데이터를 가져올수 있기 때문입니다.
반면에 규모가 커지면 service에서 service도 참조해서 데이터가 분산되지 않도록 최대한 관리할 것 같습니다. 이때, B서비스에서 A라는 서비스를 위한 메서드를 여는 것인지를 고민하면서 사용할 것 같습니다.
